GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs

```diff
@@ -865,7 +873,36 @@ protected override void OnMouseWheel(MouseEventArgs e)
             }
             else
             {
-                base.OnMouseWheel(e);
+                // Reset unconsumed wheel delta when the mouse wheel is idle.
+                if (_lastMouseWheel.ElapsedMilliseconds > 1500)
+                {
+                    _mouseWheelDeltaRemainder = 0;
+                }
+
+                int scrollLines = SystemInformation.MouseWheelScrollLines switch
+                {
+                    // Value of -1 indicates the "One screen at a time" mouse option.
+                    -1 => _rowHeight > 0 ? (Height / _rowHeight) : 1,
+                    > 0 => SystemInformation.MouseWheelScrollLines,
+                    _ => 1
+                };
+
+                int totalWheelDelta = (scrollLines * e.Delta) + _mouseWheelDeltaRemainder;
+                int wheelDeltaPerRow = SystemInformation.MouseWheelScrollDelta;
+
+                // The total wheel delta is consumed in multiples of wheelDeltaPerRow.
+                // Save the remainder and credit it to the total wheel delta during the next MouseWheel event.
+                _mouseWheelDeltaRemainder = totalWheelDelta % wheelDeltaPerRow;
+                int rowDelta = -(totalWheelDelta - _mouseWheelDeltaRemainder) / wheelDeltaPerRow;
+                if (rowDelta != 0)
+                {
+                    int toRowIndex = Math.Clamp(FirstDisplayedScrollingRowIndex + rowDelta, 0, maxRowIndex);
+
+                    // This will raise the Scroll event.
+                    FirstDisplayedScrollingRowIndex = toRowIndex;
+                }
+
+                _lastMouseWheel.Restart();
```

Comment:
So, if I open a repo, interact with it for a little while and then go out for lunch or go home, the stopwatch will continue counting... right? Is this desired? Why not just capture a timestamp at the time we exit the method, and calculate the delta when we enter the method again, and based on this delta decide whether we need to reset `_mouseWheelDeltaRemainder`?

Reply:
Why calculate `ElapsedMilliseconds` manually? Is a running `Stopwatch` expensive? I guess it just remembers its start time, and hides the fuzz with the correct handling of the elapsed time.

Reply:
> So, if I open a repo, interact with it for a little while and then go out for lunch or go home, the stopwatch will continue counting... right?

You make it sound like leaving the stove on :-D. But I share your concern and looked it up:

> Is a running Stopwatch expensive? I guess it just remembers its start time, and hides the fuzz with the correct handling of the elapsed time.

Exactly right. The `Stopwatch` class is very lightweight. It just stores a timestamp and then calculates the difference, see: [https://github.com/dotnet/.../Stopwatch.cs](https://github.com/dotnet/runtime/blob/bff0f8487965a630a8c3fa107e17ec8e02e9b32c/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs)

The downside is that `Stopwatch` uses `QueryPerformanceFrequency()` to provide high precision, which isn't needed in this case. I just benchmarked `Stopwatch.ElapsedMilliseconds` against `Environment.TickCount64` (which provides ~15ms resolution) and found the latter to be ~6x faster  on my machine (27µs vs. 4.3µs). As this runs on the UI thread and every bit counts (I think?), I could change it like this:

```
            else
            {
                // Reset unconsumed wheel delta when the mouse wheel is idle.
                long currentTickCount = Environment.TickCount64;
                if (currentTickCount - _lastMouseWheelTickCount > 1500)
                {
                    _mouseWheelDeltaRemainder = 0;
                }

                _lastMouseWheelTickCount = currentTickCount;

                int scrollLines = SystemInformation.MouseWheelScrollLines switch
                {
```

Reply:
Thank you. I confused it with S.W.F.Timer.

> just benchmarked `Stopwatch.ElapsedMilliseconds` against `Environment.TickCount64` (which provides ~15ms resolution) and found the latter to be ~7x faster on my machine (24µs vs. 3.4µs). As this runs on the UI thread and every bit counts (I think?), I could change it like this:

🚀

Reply:
Hunting µs is something that can make a difference for the revision update loop, this is executed at GUI updates so not making a difference. But keep the change.

The only part that makes sense to optimize is the revision parsing and update loop as well as the rendering in addition to which Git commands that run.. Everything else will not be seen.
(I spent some days on optimizing some regex to sort the branches and tags in the sidepanel to get it down to a few ms instead of just below 100 ms, that was in retrospect not the best spent time, but it was kind of fun.)

Reply:
I was not planning on "hunting µs" here (sounds dangerous :sunglasses:), but I try to be open for every concern and suggestion, so I wanted to offer an alternative.

Thanks for the insight!